syntax = "proto3";

package python_example.users;
option java_package = "user.generated";
option java_multiple_files = true;

message GetUserRequest {
}

message CreateUserRequest {
}

message DeleteUserRequest {
}

message AddUserCreditRequest {
  int32 amount = 1;
}

message SubtractUserCreditRequest {
  int32 amount = 1;
}

message UserRequest {
  string request_id = 1;
  string user_id = 2;

  oneof message {
    GetUserRequest get_user = 3;
    CreateUserRequest create_user = 4;
    DeleteUserRequest delete_user = 5;
    AddUserCreditRequest add_user_credit = 6;
    SubtractUserCreditRequest subtract_user_credit = 7;
  }
}

message UserResponse {
  string request_id = 1;
  string user_id = 2;
  int32 status_code = 3;

  oneof payload {
    string string = 4;
    UserState state = 5;
  }
}

message UserState {
  bool deleted = 1;
  int32 credit = 2;
}